EasyClangComplete.py

```diff
@@ -37,8 +37,8 @@
 NoneSublimeProgressStatus = progress_status.NoneSublimeProgressStatus
 PosStatus = tools.PosStatus
 CMakeFile = flags_sources.cmake_file.CMakeFile
-CMakeFileCache = flags_sources.cmake_file.CMakeFileCache
-ThreadPool = thread_pool.ThreadPool
+# CMakeFileCache = flags_sources.cmake_file.CMakeFileCache
+# ThreadPool = thread_pool.ThreadPool
```

Comment:
To be honest I don't see why commenting this lines should help us in any way. This just creates an alias (or a new reference) to the same class, right? It should never ever create a copy or am I wrong?

Reply:
Hi. This is the tricky part of python. As I mentioned before: "The original problem is due to reload modules. Every time they are reloaded the singleton will generate new instance. Thus in general we should avoid from xx import aa, we should always use the module.singleton."

Please always keep in your mind "python is dynamic script with namespace". This is completely different from java and C++. I suggest you test this behavior with some simple example code with re-import mechanism. Say we have module xx where it has class aa. When you reload xx the `xx.aa` will always point to the new instance. However, if you did `aa=xx.aa `. Reloading xx will result the aa point to different instance than `xx.aa`.

Here is my example:
```
# reloadissue.py
import random
module_spec_var = random.random() # every reload will have diffrent value

def singleton(class_):
    """Singleton class wrapper.

    Args:
      class_ (Class): Class to wrap.

    Returns:
      class_: unique instance of object.
    """
    instances = {}

    def getinstance(*args, **kwargs):
        """Get instance of a class."""
        if class_ not in instances:
            instances[class_] = class_(*args, **kwargs)
        return instances[class_]
    return getinstance

@singleton
class test_singleton(dict):
    pass


class aa(object):
    x = module_spec_var

    @classmethod
    def get_x(cls):
        return cls.x

```
Then we use this module in:
```
# show_reloadissue.py
import importlib
import random
import reloadissue
aa = reloadissue.aa
singleton0 = reloadissue.test_singleton

importlib.reload(reloadissue)
def add_to_singleton0():
    _d = singleton0()
    _d[random.random()]= random.random()

def add_to_singleton():
    _d = reloadissue.test_singleton()
    _d[random.random()]= random.random()

def add_to_reloaded_singleton():
    singleton1=reloadissue.test_singleton # this demo the singleton_cls=reloadissue.test_singleton in other module
    _d = singleton1()
    _d[random.random()] = random.random()

def singleton_issue():
    add_to_singleton0()
    add_to_singleton()
    add_to_reloaded_singleton()

    print("singleton0 this module scope only: id={}, items={}".format(id(singleton0()), singleton0()))
    print("singleton entire package scope: id={}, items={}".format(id(reloadissue.test_singleton()), reloadissue.test_singleton()))



def show_dynamic_feature():
    print("aa: id(aa)={}, aa.get_x()={}".format(id(aa), aa.get_x()))
    print("reloadissue.aa: id(aa)={}, aa.get_x()={}".format(id(reloadissue.aa), reloadissue.aa.get_x()))



if __name__ == '__main__':
    show_dynamic_feature()
    singleton_issue()
```

The result is this:
```
aa: id(aa)=94340467295720, aa.get_x()=0.47835347167596576
reloadissue.aa: id(aa)=94340467432776, aa.get_x()=0.019992688126633684
singleton0 this module scope only: id=140226399142576, items={0.2596397797967157: 0.253640595828594}
singleton entire package scope: id=140226399142664, items={0.7284843650108537: 0.21772360726146633, 0.753419975421231: 0.7037513247746104}

```
You can see, if you use `singleton0 = reloadissue.test_singleton` The singleton feature is broken cross modules. But the `reloadissue.test_singleton()` always kept its uniqueness. This is because after reload

Reply:
Ok, I get your point. But it seems that a proper solution would be to not reload any modules with a singleton in them, right?

Reply:
I think it is not easy to impose that restriction. That is why I do not like cross-module singleton in python.

Reply:
I see no way to avoid singletons. I need guarantees that the caches only exist once. This means I will try to make sure to create a module `singleton` that will never be reloaded. Maybe even write a script that would enforce that all classes marked as singletons are only declared there.